### PR TITLE
feature: support pouch ps filter

### PR DESCRIPTION
**Pull request by Ace-Tang**
support pouch ps --filter, not only equal condition, but also
unequal condition, like pouch ps --filter label=a!=b

Fixes: #1562.

Signed-off-by: Ace-Tang <aceapril@126.com>

### Ⅱ. Does this pull request fix one issue?

fixes: #1562

### CODE REVIEW

File: daemon/mgr/container_list.go [hunk 1]
```diff
@@ -0,0 +1,197 @@
...
+		case statusFilter:
+			match = fc.matchFilter(statusFilter, string(c.State.Status))
+		}
+
+		if match {
+			if name != statusFilter {
+				// consider if need to add non-running container
+				match = isNonStop || fc.all
+			}
+			break
```
Comment (yyb196) [inline daemon/mgr/container_list.go]: one match condition will stop this filter routine. what if people need both `-f label=site=daily` and `-f label=network!=overlay`

File: test/cli_ps_test.go [hunk 2]
```diff
@@ -71,6 +71,124 @@ func (suite *PouchPsSuite) TestPsWorks(c *check.C) {
 
 }
 
+// TestPsFilterEqual tests "pouch ps -f" filter equal condition work
+func (suite *PouchPsSuite) TestPsFilterEqual(c *check.C) {
+	labelA := "equal-label-a"
+	command.PouchRun("run", "-d", "--name", labelA, "-l", "a=b", busyboxImage, "top").Assert(c, icmd.Success)
+	defer DelContainerForceMultyTime(c, labelA)
+	output := command.PouchRun("inspect", labelA).Stdout()
```
Comment (Letty5411) [inline test/cli_ps_test.go]: +	output := command.PouchRun("inspect", labelA).Assert(c,icmd.Success).Stdout()

File: test/cli_ps_test.go [hunk 3]
```diff
@@ -71,6 +71,124 @@ func (suite *PouchPsSuite) TestPsWorks(c *check.C) {
...
+	_, exist1 = kv[labelA]
+	_, exist2 = kv[labelB]
+	_, exist3 = kv[name]
+	_, exist4 = kv[labelC]
+	c.Assert(exist1, check.Equals, false)
+	c.Assert(exist2, check.Equals, false)
+	c.Assert(exist3, check.Equals, true)
+	c.Assert(exist4, check.Equals, true)
+}
+
```
Comment (Letty5411) [inline test/cli_ps_test.go]: could also add negative tests, like:
-f  null
-f wrong format

and use multiple -f

File: daemon/mgr/container_list.go (hunk 1 above)
Comment (Ace-Tang) [reply daemon/mgr/container_list.go]: Yes, one match will break loop, the function used to filter the match thing, so one condition match means it is chosed, 
it is what we want, we just want output matched thing.

File: test/cli_ps_test.go (hunk 3 above)
Comment (Ace-Tang) [reply test/cli_ps_test.go]: All case has add.

File: daemon/mgr/container_list.go (hunk 1 above)
Comment (Ace-Tang) [reply daemon/mgr/container_list.go]: Modify the logic, jump until condition not equal, and add flag `jump` to determine if a condition is useful. Test case has add for this case.

File: daemon/mgr/container_list.go [hunk 4]
```diff
@@ -0,0 +1,211 @@
...
+
+	if len(fc.condition) == 0 {
+		// if no filter condition, skip filter process,
+		// add container according to flag `all` or running status
+		return isNonStop || fc.all
+	}
+
+	// filter must meets all condition
+	var match, jump, statusKey = true, true, false
+	for name := range fc.condition {
```
Comment (fuweid) [inline daemon/mgr/container_list.go]: I have idea about this part.

The `jump` boolean means we should start next round check, right?  the `matchKVFilter` will return one boolean value if we run the `filters, exist := fc.condition[field]` before check.  Therefore, we don't need to check the `jump` in the following condition. 

Does it make senses to you?

File: test/cli_ps_test.go [hunk 5]
```diff
@@ -71,6 +72,139 @@ func (suite *PouchPsSuite) TestPsWorks(c *check.C) {
 
 }
 
+// TestPsFilterInvalid tests "pouch ps -f" invalid
+func (suite *PouchPsSuite) TestPsFilterInvalid(c *check.C) {
+	result := command.PouchRun("ps", "-f", "foo")
```
Comment (fuweid) [inline test/cli_ps_test.go]: Next time, we can use the table-driven test to repeat the same thing.

File: test/cli_ps_test.go [hunk 6]
```diff
@@ -71,6 +72,139 @@ func (suite *PouchPsSuite) TestPsWorks(c *check.C) {
...
+func (suite *PouchPsSuite) TestPsFilterEqual(c *check.C) {
+	labelA := "equal-label-a"
+	command.PouchRun("run", "-d", "--name", labelA, "-l", "a=b", busyboxImage, "top").Assert(c, icmd.Success)
+	defer DelContainerForceMultyTime(c, labelA)
+	output := command.PouchRun("inspect", labelA).Assert(c, icmd.Success).Stdout()
+	result := []types.ContainerJSON{}
+	if err := json.Unmarshal([]byte(output), &result); err != nil {
+		c.Errorf("failed to decode inspect output: %v", err)
+	}
+	labelAName := result[0].ID
```
Comment (fuweid) [inline test/cli_ps_test.go]: s/labelAName/labelAID/ maybe better

File: apis/swagger.yml [hunk 7]
```diff
@@ -3529,6 +3529,33 @@ definitions:
...
 
+  ContainerListOptions:
+    description: |
+      options of list container, filters (a `map[string][]string`) to process on the container list. Available filters:
+
+      - `id=container-id`
+      - `name=container-name`
+      - `status=running`
+      - `label=key` or `label=key=value`
+      - `network=container-network`
```
[the remark this answers is not in the document]
Comment (Ace-Tang) [reply apis/swagger.yml]: I do not think so, interface should think more, we just tell which not support before this, user always not care about defined field

File: cli/ps.go [hunk 8]
```diff
@@ -50,15 +52,26 @@ func (p *PsCommand) addFlags() {
...
+	filter, err := opts.ParseFilter(p.flagFilter)
+	if err != nil {
+		return err
+	}
+
 	var containers containerList
-	containers, err := apiClient.ContainerList(ctx, p.flagAll)
+
+	option := types.ContainerListOptions{All: p.flagAll}
+	if len(filter) > 0 {
```
[the remark this answers is not in the document]
Comment (Ace-Tang) [reply cli/ps.go]: No, less function call is better, `api.Client.ContainerList` has `len(filter) > 0` for robust.

File: client/container_list_test.go [hunk 9]
```diff
@@ -57,7 +59,9 @@ func TestContainerList(t *testing.T) {
 	client := &APIClient{
 		HTTPCli: httpClient,
 	}
-	containers, err := client.ContainerList(context.Background(), true)
+	containers, err := client.ContainerList(context.Background(), types.ContainerListOptions{
+		All: true,
```
[the remark this answers is not in the document]
Comment (Ace-Tang) [reply client/container_list_test.go]: Construct `Container` struct is complete, let ci do this filter check.

File: daemon/mgr/container_list.go [hunk 10]
```diff
@@ -0,0 +1,211 @@
...
+import (
+	"context"
+	"fmt"
+	"regexp"
+	"strings"
+
+	"github.com/alibaba/pouch/apis/opts"
+)
+
+var (
```
[the remark this answers is not in the document]
Comment (Ace-Tang) [reply daemon/mgr/container_list.go]: Not following you, could you explain more detail? @fuweid

File: daemon/mgr/container_list.go [hunk 11]
```diff
@@ -0,0 +1,211 @@
...
+	return false, match
+}
+
+// filter does all select container work.
+func (fc *filterContext) filter(c *Container) bool {
+	isNonStop := c.IsRunningOrPaused()
+
+	// if filterFunc is defined, we not try other filter condition
+	if fc.filterFunc != nil {
+		if match := fc.filterFunc(c); match {
```
[the remark this answers is not in the document]
Comment (Ace-Tang) [reply daemon/mgr/container_list.go]: good

File: daemon/mgr/container_list.go (hunk 4 above)
Comment (Ace-Tang) [reply daemon/mgr/container_list.go]: I also think about this, `filters, exist := fc.condition[field]` if field not exist, whether filter function return true or false is not make sense. Maybe need more think.

File: test/cli_ps_test.go (hunk 5 above)
Comment (Ace-Tang) [reply test/cli_ps_test.go]: ^-^

File: test/cli_ps_test.go (hunk 6 above)
Comment (Ace-Tang) [reply test/cli_ps_test.go]: This rule is for variable, WDYT?

File: apis/opts/filter.go [hunk 12]
```diff
@@ -0,0 +1,105 @@
...
+	rawdata, err := json.Marshal(filter)
+	if err != nil {
+		return "", fmt.Errorf("failed to marshal filter params %v", filter)
+	}
+
+	return string(rawdata), nil
+}
+
+// FromURLParam gets filter from url query
+func FromURLParam(param string) (map[string][]string, error) {
```
[the remark this answers is not in the document]
Comment (Ace-Tang) [reply apis/opts/filter.go]: Oh, good, I think about move the file `api/opts/filter.go` to `pkg` package and make it a package named filter, and I think the name `FromURLParam` and `ToURLParam` will make sense.

File: daemon/mgr/container_list.go [hunk 13]
```diff
@@ -0,0 +1,211 @@
...
+
+	if len(equalKV) == 0 && len(unequalKV) == 0 {
+		return false, false
+	}
+
+	match := false
+	for k, v := range value {
+		// filter equal condition
+		if equalValue, exist := equalKV[k]; exist {
+			if equalValue == "" {
```
[the remark this answers is not in the document]
Comment (Ace-Tang) [reply daemon/mgr/container_list.go]: good

File: client/container_list_test.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (fuweid) [reply client/container_list_test.go]: 😯 you are right.

File: apis/swagger.yml (hunk 7 above)
[the remark this answers is not in the document]
Comment (fuweid) [reply apis/swagger.yml]: Solved

**Comment by YaoZengzeng** [comment]
LGTM, it's all right for CRI part :)

**Pull request merged**